src/disco/shred/fd_shred_tile.c

```diff
@@ -0,0 +1,694 @@
+
+  /* Reload stake contact info if it has changed */
+  /* FIXME: Be careful when we do this to make sure we don't get
+     data for the wrong epoch. */
+  fd_mvcc_t * cluster_nodes_mvcc = ctx->args->cluster_nodes_mvcc;
+
+  ulong version_a = fd_mvcc_version_query( cluster_nodes_mvcc );
+  if( FD_LIKELY( !(version_a % 2) & (ctx->prev_contact_version != version_a) ) ) {
+    FD_LOG_NOTICE(( "reloading contact info" ));
+    int cluster_nodes_updated = 1;
+    for(;;) {
+      version_a = fd_mvcc_version_query( cluster_nodes_mvcc );
+      if( FD_UNLIKELY( version_a % 2 ) ) {
+        /* writer started writing. Bail and try again later */
+        cluster_nodes_updated = 0;
+        break;
+      }
+
+      uchar const * mvcc_app = fd_mvcc_app_laddr_const( cluster_nodes_mvcc );
+      ulong dest_cnt     = ((ulong const *)fd_type_pun_const( mvcc_app ))[0];
+      ulong total_weight = ((ulong const *)fd_type_pun_const( mvcc_app ))[1];
+      ulong slot_start   = ((ulong const *)fd_type_pun_const( mvcc_app ))[2];
+      ulong slot_cnt     = ((ulong const *)fd_type_pun_const( mvcc_app ))[3];
+      ulong epoch        = ((ulong const *)fd_type_pun_const( mvcc_app ))[4];
+      /* TODO: Handle overflow case by making an entry with
+         remaining weight at the end */
+      (void)total_weight;
+
+      FD_TEST( slot_cnt );
+      FD_TEST( dest_cnt );
+      FD_TEST( dest_cnt < MAX_SHRED_DESTS );
+
+      fd_shred_dest_wire_t const * in_dests = fd_type_pun_const( mvcc_app + 5UL*sizeof(ulong) );
+
+      ulong staked_cnt = 0UL;
+      for( ulong i=0UL; i<dest_cnt; i++ ) {
+        /* Resolve the destination */
+        int can_send_to_dest = 0;
+        if( FD_LIKELY( in_dests[i].ip4_addr ) ) {
+
+          uint out_next_ip[1];
+          uint out_ifindex[1];
+          int res = fd_ip_route_ip_addr( ctx->shred_dest[i].mac_addr, out_next_ip, out_ifindex, ctx->args->ip, in_dests[i].ip4_addr );
+
+          if( FD_LIKELY( res==FD_IP_SUCCESS ) ) {
+            can_send_to_dest = 1;
+            ctx->shred_dest[i].ip4            = in_dests[i].ip4_addr;
+            ctx->shred_dest[i].port           = in_dests[i].udp_port;
+          }
+          else if( FD_LIKELY( res==FD_IP_PROBE_RQD ) ) {
+            ulong tsorig = fd_frag_meta_ts_comp( fd_tickcount() );
+
+            uchar * arp_packet = fd_chunk_to_laddr( ctx->net_wksp, ctx->net_chunk );
+            ulong arp_sz[1];
+            res = fd_ip_arp_gen_arp_probe( arp_packet, sizeof(fd_ip_arp_t), arp_sz, *out_next_ip, ctx->args->src_ip, ctx->args->src_mac );
+            if( res!=FD_IP_SUCCESS ) FD_LOG_ERR(( "Generation of arp probe failed" ));
+
+            ulong tspub = fd_frag_meta_ts_comp( fd_tickcount() );
+            ulong   ctl = fd_frag_meta_ctl( ctx->tile_idx, 1, 1, 0 );
+            ulong   sig = fd_disco_netmux_sig( 0U, 0U, SRC_TILE_SHRED, SRC_TILE_NET ); /* arp is not IP */
+            fd_mcache_publish( ctx->args->to_net_mcache, ctx->net_depth, ctx->net_seq, sig, ctx->net_chunk,
```

Comment:
I don't think it's valid to send the probe out here? You need to wait until the MVCC read has completed and is known valid, else it might be torn.

Reply:
Good catch. I added a check that the read hasn't been torn before sending out the probe.